### ISSUE TITLE
Estimator does not not support sparse jobs

### ISSUE THREAD

**Issue by yaroslavvb** [tensorflow/tensorflow#14502]
You need sparse jobs (https://github.com/tensorflow/tensorflow/commit/8177edd7700ccbe0c831e680a1acb5275819d762) if you want async training to proceed without having all workers available.

This currently doesn't work because estimator [uses json](https://github.com/tensorflow/tensorflow/blob/9ff1aebc2b3dcb07f15b92c0a58027b0438f502a/tensorflow/python/estimator/run_config.py#L379) to dump values into `TF_CONFIG` env var, and json does not allow numeric keys. `json.dumps` automatically converts numeric keys into strings.

However, clusterspec for sparse job must have integers like `{"local": {37: "localhost:0"}}` and will crash if we have `"37"` instead of `37`.

Suggestion 1: Modify sparse job support to allow strings as task indices

Suggestion 2: use pickle + base16 encoding to transmit these values

Base16 is shell-friendly, it allows you to copy paste the value into "export TF_...." in terminal

```
 # saving
 cluster_config = {'cluster': cluster_spec, 'task': task_spec}
 pickle_string = pickle.dumps(sparse_cluster_config)
 pickle_string_encoded = base64.b16encode(pickle_string)
 pickle_string_encoded = pickle_string_encoded.decode('ascii')
 export_command = "export TF_PICKLE_BASE16=%s"%(pickle_string_encoded,)
 os.system(export_command)

# loading
config_dict = pickle.loads(base64.b16decode(os.environ["TF_PICKLE_BASE16"]))
config.task_type = config_dict["task"]["type"]
config.task_id = config_dict["task"]["index"]
config.cluster_spec = config_dict["cluster"]
return config
```

cc @ispirmustafa

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and this issue has an assignee.Please update the label and/or status accordingly.

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and this issue has an assignee.Please update the label and/or status accordingly.

**Comment by yaroslavvb** [comment]
PS, this is one reason why Estimator sucks for large training jobs -- dense job spec means training can't start until all the workers are up

**Comment by xiejw** [comment]
@martinwicke for priority.

**Comment by xiejw** [comment]
Hi yaroslavvb, thanks for tracking this. 

As the TF_CONFIG contains all the information required to populate the RunConfig properties, we need to be careful about how to support this. For example, the num_worker_replicas is calculated based on the number of the training jobs inside the TF_CONFIG. If we allow sparse cluster spec directly at environment variable TF_CONFIG level, num_workers_replicas will be filled with wrong value. User will be very hard to automatically figure out how to configure the SyncReplicaOptimizer, which replies on the number of the worker in the system. 

One alternative option could be: We keep TF_CONFIG as is, but use an opt-in approach to turn on the sparse cluster spec when bring up the tf.train.Server. To be precise, at Estimator level, we still see the global information for the whole cluster, but when launching the tf.train.Server to connect to each other, we take the advantage of the sparse cluster spec. 

What's your thoughts on this? Are you willing to contribute here? 

If yes, we could work on a plan how to change/test here.

**Comment by yaroslavvb** [comment]
@xiejw not sure about contributing here because I'm not sure whether Estimator is going to fit this use-case. IE, I need high performance models at large scale and the official high performance models [guide](https://www.tensorflow.org/performance/performance_models) doesn't use Estimator. Also the advice on this list has been to use that instead of Estimator.

**Comment by xiejw** [comment]
So, @yaroslavvb, then what's your proposal here? Maybe I misread your comment, but feel like you do not need this feature, right?

**Comment by yaroslavvb** [comment]
I guess I'm saying that it makes sense to implement this feature as part of concerted effort to make Estimator work on large-scale problems. I didn't want to do it since I don't know what Estimator plans are.

**Comment by xiejw** [comment]
Given this, I am going to close this and revisit/reopen once we have use case or clients. Maybe very soon.
